optuna/storages/_journal/redis.py

```diff
@@ -0,0 +1,62 @@
+import json
+from typing import Any
+from typing import Dict
+from typing import List
+
+from optuna._experimental import experimental_class
+from optuna._imports import try_import
+from optuna.storages._journal.base import BaseJournalLogStorage
+
+
+with try_import() as _imports:
+    import redis
+
+
+@experimental_class("3.1.0")
+class JournalRedisStorage(BaseJournalLogStorage):
+    """Redis storage class for Journal log backend.
+
+    Args:
+        url:
+            URL of the redis storage, password and db are optional. (ie: redis://localhost:6379)
+
+    """
+
+    def __init__(self, url: str) -> None:
+
+        _imports.check()
+
+        self._url = url
+        self._redis = redis.Redis.from_url(url)
+
+    def read_logs(self, log_number_from: int) -> List[Dict[str, Any]]:
+
+        self._redis.setnx("log_number", -1)
+        max_log_number_bytes = self._redis.get("log_number")
+        assert max_log_number_bytes is not None
+        max_log_number = int(max_log_number_bytes)
+
+        logs = []
+        last_decode_error = None
+        for log_number in range(log_number_from, max_log_number + 1):
+            if last_decode_error is not None:
+                raise last_decode_error
+            log_bytes = self._redis.get(self._key_log_id(log_number))
+            assert log_bytes is not None
+            log = log_bytes.decode("utf-8")
+            try:
+                logs.append(json.loads(log))
+            except json.JSONDecodeError as err:
+                last_decode_error = err
+        return logs
```

Comment:
I have two suggestions.

1. As not like JournalFileStorage, we can remove `last_decode_error` variable since we've already obtained `max_log_number`.
2. `self._redis.get(self._key_log_id(log_number))` may return `None` when running with multiple workers. So how about retrying it?

```suggestion
        logs = []
        for log_number in range(log_number_from, max_log_number + 1):
            log_bytes = self._redis.get(self._key_log_id(log_number))
            if log_bytes is None:
                for i in range(10):
                    time.sleep(min(0.05 * (2 ** i), 10))
                    log_bytes = self._redis.get(self._key_log_id(log_number))
                    if log_bytes is not None:
                        break
                else:
                    raise Exception("retry error...")  # TODO: Improve the error message.
            log = log_bytes.decode("utf-8")
            try:
                logs.append(json.loads(log))
            except json.JSONDecodeError as err:
                if log_number != max_log_number:
                    raise err
        return logs
```

The exponential backoff intervals of my suggested code are like the following:

```
>>> for i in range(10):
...     print(min(0.05 * (2 ** i), 10))
...
0.05
0.1
0.2
0.4
0.8
1.6
3.2
6.4
10
10
```